Remove Haskell, Add Rust.

Codysseus opened the pull request:
This commit removes all Haskell files and adds files important
for Cargo, as well as two files in Rust: main.rs, syntax.rs.

I have a basic working Syntax that I tried to make similar to
the previous abstract syntax with some improvements. I removed
the different types of comments and simply made it a vector of
strings. I also made Function definitions and Shader definiti-
ons part of the Expression syntax.

Notice too that the Block expression is simply a vector of
Boxed expressions. This is something I don't quite understand
yet but it is what I saw other people doing in their AST
implementations.

Please refer to this link for my source of inspiration:
https://createlang.rs/01_calculator/ast.html

This is in reference to issue #14

Codysseus committed:
Remove Haskell, Add Rust.

This commit removes all Haskell files and adds files important
for Cargo, as well as two files in Rust: main.rs, syntax.rs.

I have a basic working Syntax that I tried to make similar to
the previous abstract syntax with some improvements. I removed
the different types of comments and simply made it a vector of
strings. I also made Function definitions and Shader definiti-
ons part of the Expression syntax.

Notice too that the Block expression is simply a vector of
Boxed expressions. This is something I don't quite understand
yet but it is what I saw other people doing in their AST
implementations.

Please refer to this link for my source of inspiration:
https://createlang.rs/01_calculator/ast.html

montymxb commented on src/syntax.rs:
```diff
@@ -0,0 +1,105 @@
+        value: Block,
+    },
+    Def {
+        mutable: bool,
```
I think it might be a better idea to introduce separate `Def` and `MutDef` exprs.

montymxb commented on src/syntax.rs:
```diff
@@ -0,0 +1,105 @@
+    Ref(String),
+    App {
+        fname: String,
+        arguments: Block,
```
Shouldn't application args be a list of `Box<Expr>`? I.e. `Vec<Box<Expr>>`

Codysseus replied:
I mean, that is what a `Block` is, but I was just reusing the type. I am actually considering removing the `Block` type since I am not sure it adds anything to this AST.

montymxb replied:
Oh right, gotcha didn't see that about `Block` the first time around.

In regards to whether it adds anything, I'm inclined to agree, but it's making me think about how a vector of exprs is really kind of a vector of statements, but not really :/....

Codysseus replied:
Weren't we taking an Imperative syntax? i.e. a list of statements with the last statement evaluating to the value of the block

montymxb replied:
We are. It was more a comment on how this feels like we're mixing Statements and Expressions together here, when they're really separate. Expressions should always resolve to final values, but statements do not necessarily do that. For example, with the way this is written now, we can write an expression ```BinOp Add 1 (Update "x" 5)```, but that's contingent on Update being a valid Expr, which would need to resolve to a final value. We *could* make Update resolve to the value of the updated name, but I don't know about that...

montymxb replied:
To clarify, in some languages statements can constitute valid expressions, like `x = y = 2`. Where the value of the statement is just the value that is assigned to a name. This kind of muddies the waters however, but it's an option we could take if we can justify a final value for all 'statements', such that they constitute valid expressions (and then we can keep everything under the category of Expr like it is now).

montymxb commented on src/syntax.rs:
```diff
@@ -0,0 +1,105 @@
+        e1: Box<Expr>,
+        e2: Box<Expr>,
+    },
+    Compose(String, String),
```
This allows composition of named functions, but not for unnamed functions (ex. lambdas). If we allow unnamed functions, we probably will need to change this.

Codysseus replied:
Very true! In the old AST, `Compose` was a `BinOp`. I think I agree with you that this should be the case.

Codysseus replied:
This also brings up the question, how are we representing composition of shaders in the AST?

montymxb replied:
Assuming functions are still first class values, it would just be a binary operation on functions as values. With the way this syntax is written, we could drop the 'name' property from `Func`, and be able to create unnamed functions that could be bound to a name via `Def` (or `MutDef`). Then I think that should work okay 🤔

montymxb commented on src/syntax.rs:
```diff
@@ -0,0 +1,105 @@
+        return_type: Type,
+        body: Block,
+    },
+    Shader {
```
Do we want to have a shader function type? I thought we were thinking about doing shaders as functions directly, and instead introducing types that those functions could use to generate the desired shader effects.

Codysseus replied:
Ah I remember now, yes. You're talking about the difference between the functions returning a type, and shaders returning named types?

montymxb replied:
Yeah. A shader is generated through a series of function applications that produce the right types for that shader. Something basic would be Vec4 for position and Vec4 for color, where Position and Color are just type synonyms for those types. We could also be more aggressive about keeping those types separate so we can distinguish the set of operations that can be performed on each, make it a little less ambiguous about whether they can be mixed (or not).

Codysseus replied:
OK so I had an idea a bit ago and I will lay it out. We want to make sure our users don't mess up their shaders, right? I was reading a bit of Ada and I like their idea of a specific section for declaration of variables. My idea is to have a specific section at the end of the function for setting variables that are returned! That way we don't have to worry about searching inside ASTs for modifications of the returned variable. This would mean that we don't have a return expression, but I don't think that preventing users from performing an early return is a bad thing.
What do you think?

montymxb reviewed (commented):
Some questions about the AST choices, but the rest seems okay so far. My rust is *rusty*, so I will need to read some more to be able to understand what these choices entail.

Codysseus committed:
Def -> Def | DefMut

Codysseus commented:
A bit of a wrench here, we originally talked about reworking the type hierarchy. Can we write up a formal definition of the type hierarchy? I feel like that is a good step since ATM we are working on the implementation, whereas we do not necessarily know /what/ we are implementing. LMK if/when you want to meet up to talk about this. I will write up my ideas and present them, then we can document the meeting and come to a conclusion.

montymxb commented:
I think that is an *excellent* idea to chat about this! As I was reading your note related to an explicit declaration section (which reminds me a lot of an 'exports' section, such as those at the end of a module in Node.js), I was thinking that maybe it would be good to see if the type system doesn't solve both of these problems for us (the notion of what to check, and the notion of the hierarchy before writing up stuff).

Codysseus committed:
Modify the Expr, Type. Remove Block, ShaderType, DiscreteType

Type now contains
	Int
	Uint
	Bool
	Float
	Double
	Array
	Struct
	Function

I don't know that we need a Tuple type, as that is just a 2
length array.

I added
	D
	Return
to Expr because we need a way to represent a double immediate
and I think that we can have a return statement. If we're
making this a stepping stone to GLSL then I think it could be
reasonable to allow that.

montymxb committed:
adds front-end under web

montymxb committed:
Merge branch 'rust-migration' of https://github.com/montymxb/FDSSL into rust-migration

montymxb commented:
Alright, the front-end has been added underneath **web** for now. I think that should suffice, and I'm good with a merge to change the project direction 👍
